### ISSUE TITLE
Conformance testing with FIDO-alliance tool

### ISSUE THREAD

**Issue by abergs** [abergs/fido2-net-lib#13]
Will update this issues with testing progress

**Comment by aseigler** [comment]
I consider "passing" to mean all tests pass the tool check with zero exceptions thrown for the expected pass tests and expected fail tests result in a thrown exception from the library. 

These set of tests should all be passing:

ServerAuthenticatorAttestationResponse-Resp-1
ServerAuthenticatorAttestationResponse-Resp-2
ServerAuthenticatorAttestationResponse-Resp-3

**Comment by abergs** [comment]
Latest run:

- ✅Server-ServerAuthenticatorAssertionResponse-Resp-1 Test server processing ServerAuthenticatorAssertionResponse structure
- ✅ Server-ServerAuthenticatorAssertionResponse-Resp-2 Test server processing CollectClientData
- ❌ Server-ServerAuthenticatorAssertionResponse-Resp-3 Test server processing authenticatorData
- ✅ Server-ServerPublicKeyCredentialGetOptionsResponse-Req-1 Test server generating ServerPublicKeyCredentialGetOptionsResponse
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-1 Test server processing ServerAuthenticatorAttestationResponse structure
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-2 Test server processing CollectClientData
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-3 Test server processing AttestationObject
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-4 Test server support of the authentication algorithms
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-5 Test server processing "packed" FULL attestation
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-6 Test server processing "packed" SELF(SURROGATE) attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-7 Test server processing "none" attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-8 Test server processing "fido-u2f" attestation
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-9 Test server processing "tpm" attestation
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-A Test server processing "android-key" attestation
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-B Test server processing "android-safetynet" attestation
- ✅ Server-ServerPublicKeyCredentialCreationOptions-Req-1 Test server generating ServerPublicKeyCredentialCreationOptionsRequest 

Note: Strange errors when running *Server-ServerAuthenticatorAssertionResponse-Resp-3*.
![image](https://user-images.githubusercontent.com/357283/44310548-97369400-a3d8-11e8-8969-5154a2e9acfa.png)

**Comment by aseigler** [comment]
I am not sure I agree with some of the test results compared to the spec.  I may open an issue or two against the conformance tool.

**Comment by aseigler** [comment]
Not mine but I'm not the only one. 

https://github.com/fido-alliance/conformance-tools-issues/issues/375

**Comment by aseigler** [comment]
A bunch more of these should be passing now. However, some of the failure tests are passing not because we are defending properly but because they are hitting other unexpected exceptions. It's tedious to track which ones are "good" pass and which are "bad" pass.

**Comment by abergs** [comment]
Latest run:
- ✅ Server-ServerPublicKeyCredentialCreationOptions-Req-1 Test server generating ServerPublicKeyCredentialCreationOptionsRequest
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-1 Test server processing ServerAuthenticatorAttestationResponse structure
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-2 Test server processing CollectClientData
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-3 Test server processing AttestationObject
- ❌(1) Server-ServerAuthenticatorAttestationResponse-Resp-4 Test server support of the authentication algorithms
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-5 Test server processing "packed" FULL attestation
- ❌ (1) Server-ServerAuthenticatorAttestationResponse-Resp-6 Test server processing "packed" SELF(SURROGATE) attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-7 Test server processing "none" attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-8 Test server processing "fido-u2f" attestation
- ❌(1) Server-ServerAuthenticatorAttestationResponse-Resp-9 Test server processing "tpm" attestation
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-A Test server processing "android-key" attestation
- ❌ Server-ServerAuthenticatorAttestationResponse-Resp-B Test server processing "android-safetynet" attestation
- ✅ Server-ServerPublicKeyCredentialGetOptionsResponse-Req-1 Test server generating ServerPublicKeyCredentialGetOptionsResponse
- ✅ Server-ServerAuthenticatorAssertionResponse-Resp-1 Test server processing ServerAuthenticatorAssertionResponse structure
- ❌ Server-ServerAuthenticatorAssertionResponse-Resp-2 Test server processing CollectClientData
- ❌ Server-ServerAuthenticatorAssertionResponse-Resp-3 Test server processing authenticatorData

I agree with you about finding "holes" in the defense where we catch another error in the response before we reach what test is supposed to test for. I guess the best way to completely solve that would be to add test cases in the conformance tool.

**Comment by aseigler** [comment]
I am of the opinion that something is broken in Server-ServerAuthenticatorAssertionResponse-Resp-3. I get the same results you do and the test does not seem to be hitting any interesting code at all.

**Comment by aseigler** [comment]
Lots more should be passing now. Most of Server-ServerAuthenticatorAssertionResponse-Resp-3 is still broken, but it seems that is a known issue now.

**Comment by abergs** [comment]
@aseigler this branch contains a commit that makes the tests pass. It's still a bit ugly and the UV/UP tests seems to fail a lot.

I pushed it to a branch because I too eagerly made a change from Raw.Id to Raw.RawId and want to fix that before merging to master but you can use it if you want to run tests.

**Comment by abergs** [comment]
I agree with the vice-versa @aseigler ;) But feel free to use the branch to run the tests since they work in that branch.

**Comment by abergs** [comment]
Latest run:
- ✅ Server-ServerPublicKeyCredentialCreationOptions-Req-1 Test server generating ServerPublicKeyCredentialCreationOptionsRequest
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-1 Test server processing ServerAuthenticatorAttestationResponse structure
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-2 Test server processing CollectClientData
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-3 Test server processing AttestationObject
- 💠 Server-ServerAuthenticatorAttestationResponse-Resp-4 Test server support of the authentication algorithms
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-5 Test server processing "packed" FULL attestation
- ❌ (1) Server-ServerAuthenticatorAttestationResponse-Resp-6 Test server processing "packed" SELF(SURROGATE) attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-7 Test server processing "none" attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-8 Test server processing "fido-u2f" attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-9 Test server processing "tpm" attestation
- 💠 Server-ServerAuthenticatorAttestationResponse-Resp-A Test server processing "android-key" attestation
- ✅ Server-ServerAuthenticatorAttestationResponse-Resp-B Test server processing "android-safetynet" attestation
- ✅ Server-ServerPublicKeyCredentialGetOptionsResponse-Req-1 Test server generating ServerPublicKeyCredentialGetOptionsResponse
- ✅ Server-ServerAuthenticatorAssertionResponse-Resp-1 Test server processing ServerAuthenticatorAssertionResponse structure
- ✅ Server-ServerAuthenticatorAssertionResponse-Resp-2 Test server processing CollectClientData
- ❌(2) Server-ServerAuthenticatorAssertionResponse-Resp-3 Test server processing authenticatorData

💠 Pending - What does this mean?
Server-ServerAuthenticatorAssertionResponse-Resp-3  fails on UV/UP flags.

**Comment by abergs** [comment]
@aseigler Do you know what the pending things are?

**Comment by aseigler** [comment]
Functionally, off the top of my head, I know about these:

1. Missing ALG_SIGN_ED25519_EDDSA_SHA512_RAW support
2. Finish parsing Android key extensions
3. https://github.com/fido-alliance/conformance-tools-issues/issues/387
4. Making use of MDS

MDS doesn't seem to have any authenticators registered with aaguid yet, but there is an update supposed to post on 18-Sept that may change that.  I am a little baffled by the MDS conformance tests, they do not seem to work like the rest of the tests which are fairly straightforward.

**Comment by abergs** [comment]
That's good. But I was as actually refering to the tests that report status as "pending" with the blue color. (so not "not passing" or "passing". just "pending")

**Comment by aseigler** [comment]
Ah, OK. Those blue marks coincide with #1 and #2 above. The first X is #3 and #4 is not referenced yet anywhere else.

**Comment by aseigler** [comment]
Android key should pass clean after tonight's PR. I don't see a good way to implement Ed25519  support without taking a dependency on something like https://github.com/CodesInChaos/Chaos.NaCl. There doesn't seem to be native support that I can find (and I looked, hard).

**Comment by aseigler** [comment]
With the metadata handling code from https://github.com/abergs/fido2-net-lib/pull/42 minus the outstanding issues https://github.com/fido-alliance/conformance-tools-issues/issues/395 and https://github.com/fido-alliance/conformance-tools-issues/issues/396, and without figuring out EdDSA support, we should be clean for all 4 authenticator assertion and attestation tests.

**Comment by aseigler** [comment]
After #43 we should be clean for all server tests aside from metadata service tests, for all algorithms except ED25519.  Please verify independently.

**Comment by abergs** [comment]
🎉 100% test passes with all optional algorithms (except ED25519)
![image](https://user-images.githubusercontent.com/357283/46612896-12cbdc00-cb12-11e8-8a72-cc564c6cdde8.png)

This calls for celebration. @aseigler an incredible well done job! 
Closing this issue now. If you think we should add support for ED25519, please open a new issue.

**Comment by aseigler** [comment]
I have a fix for EdDSA support I am testing here https://github.com/aseigler/fido2-net-lib/commit/e61e8fb3b8b0b40ccfe2f19208d108969a3d64e6 and here https://github.com/aseigler/fido2-net-lib/commit/5743c4b2f85fa4f85be38d31a7e198e597a3627a.

**Comment by pham-vy** [comment]
Need help with Metadata Service tests
At now the server does not check certificate revocation, so the F-5 (intermediate certificate is revoked) and F-6 (leaf certificate is revoked) in Metadata Service tests are failed.
But if I change the Revocation Mode to Online, the chain Status will be Revocation Status Unknown and Offline Revocation. It seems like no CRL distribution point in MDSROOT.crt
How can I get the CRL to check revocation in these test cases?

**Comment by aseigler** [comment]
There are links to the root and issuing CRLs here: https://fidoalliance.org/metadata/

But there is no in box way to use CRL file with X509Certificate2, and I haven't gotten the cycles to write out something that will fill the gap.

**Comment by pham-vy** [comment]
How can I pass the case:
F-2 Send ServerAuthenticatorAttestationResponse with SELF "packed" attestation, that contains full attestation, and check that server returns an error

Already known that I have to use metadata for this case, but there is no suitable entry from metadata.

**Comment by aseigler** [comment]
You have to use the test metadata from the link in the conformance tool. See the custom metadata load stuff in the metadata code and the packed attestation code that interacts with metadata to see how that test was covered.

**Comment by aseigler** [comment]
I finally figured out how to pass ALL of the conformance tests, including the metadata service tests.  You get a snazzy "submit results" button! I will document the metadata service test pass process so that it can be repeated by anyone.  You do NOT need an MDS access key to pass the metadata service tests.

![image](https://user-images.githubusercontent.com/6605560/51702232-42c7be00-1fe1-11e9-9a1b-f02f6cd2a99b.png)

**Comment by idaviddesmet** [comment]
Really great news! 😃
